### PR TITLE
[Structured-data-generator] adding structured_data_generator function

### PR DESCRIPTION
**Pull request by ZeevRispler**

### CODE REVIEW

File: json_data_generator/json_data_generator.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                current_chunk_size = chunk_size
+                amount -= chunk_size
+            else:
+                current_chunk_size = amount
+
+            # Create the prompt:
+            prompt = prompt_structure.format(
+                amount=current_chunk_size,
+            )
+            # Generate the conversation:
```
Comment (guy1992l) [inline json_data_generator/json_data_generator.py:85]: Generate a chunk of data

File: json_data_generator/json_data_generator.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            chunk_data = llm.predict(text=prompt)
+            chunk_data = chunk_data[chunk_data.find("["):chunk_data.rfind("]") + 1]
```
Comment (guy1992l) [inline json_data_generator/json_data_generator.py:87]:
```suggestion
            # Validate the response for correct python `list` structure
            chunk_data = chunk_data[chunk_data.find("["):chunk_data.rfind("]") + 1]
```

File: json_data_generator/json_data_generator.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    # Load the OpenAI model using langchain:
+    os.environ["OPENAI_API_KEY"] = _env_or_secret(context, key="OPENAI_API_KEY")
+    os.environ["OPENAI_API_BASE"] = _env_or_secret(context, key="OPENAI_API_BASE")
+    llm = ChatOpenAI(model=model_name)
+
+    # Start generating data:
+    data = []
+    for _ in tqdm.tqdm(range(int(amount / chunk_size) + 1), desc="Generating"):
+        for tryout in range(3):
+            if amount > chunk_size:
```
Comment (guy1992l) [inline json_data_generator/json_data_generator.py:75]: Get the size

File: json_data_generator/json_data_generator.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            chunk_data = chunk_data[chunk_data.find("["):chunk_data.rfind("]") + 1]
+            if chunk_data.count("[") != chunk_data.count("]"):
+                print("Failed to get proper json format from model, number of '[' doesn't match number of ']'.")
+                continue
+            chunk_data = ast.literal_eval(chunk_data)
+            data += chunk_data
+            break
+        if tryout == 3:
+            raise ValueError(
+                f"Could not generate a proper json format for the given fields, using given model: {model_name}. Hint: Gpt-4 works for our purpose.")
```
Comment (guy1992l) [inline json_data_generator/json_data_generator.py:96]:
```suggestion
                f"Could not generate a proper json format for the given fields, using given model: {model_name}. Hint: Gpt-4 works best for most scenarios.")
```

**Review by guy1992l — CHANGES_REQUESTED** [review]
The `json` in `json_data_generator` is kinda misleading. If the user is not using MLRun then he won't receive a json, but a `list`. I think the function name should be changed to something like structured data generator or something else if you find a better name. The json part is being done via the `returns` where you specified you want to log the returned `list` as a json file.

Also, don't forget the section comments (`#`) :)

File: json_data_generator/function.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+metadata:
+  name: json-data-generator
+  tag: ''
+  hash: 1d12e2f61c526d24b384d2c80315839f26a82c0a
+  project: ''
+  labels:
+    author: zeevr
+  categories:
+  - machine-learning
+  - Data Preparation
```
[the remark this answers is not in the document]
Comment (ZeevRispler) [reply json_data_generator/function.yaml:11]: i think we can do both, i was trying to fit to existing categories

File: json_data_generator/json_data_generator.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import os
+import tqdm
+from langchain.chat_models import ChatOpenAI
+import ast
+
+
+def _env_or_secret(context, key):
+    if key in os.environ:
+        return os.environ[key]
+    return context.get_secret(key)
```
[the remark this answers is not in the document]
Comment (ZeevRispler) [reply json_data_generator/json_data_generator.py:24]: but you said remove all mlrun context, why use it here?

**Review by guy1992l — CHANGES_REQUESTED** [review]
minor stuff, the logic and function itself is sweet 💪 Please make sure to use `black` and `isort` to format the code before pushing.

**Pull request merged**